### PR TITLE
Add class-to-module project manifest

### LINKED ISSUE #1171: [CHANGE] - Make system know, which classes belong to which module.
### Summary

We currently can have the situation, that classes are not known during project loading, if modules are missing.

We though don't have any information, which class belongs to which module. This might be useful though, if modules provided data classes without a package.

The idea:
 - Make factories remember the modules of each registered class (the module interface does not need to be changed imho)
 - During Serialization, add module ID to each object as an attribute (this though increases the size of the projects!)

### Impact on the codebase

- Serialization
- Factories

### Current workarounds

None

### PR DESCRIPTION
**Pull request by rainman110**
## Description

  Implements a class-to-module manifest for GTlab projects, resolving #1171.

  Object factories now retain the module ID associated with each registered class. Projects serialize this information under the global `METADATA/CLASS-PROVIDERS` node, allowing GTlab to determine which  modules provide the classes used by a project.

  Class-provider metadata is stored in:

  - The main `project.gtlab` file
  - Externalized object files
  - Module package `.gtmod files`, filtered to classes used by the package
  - Task `.gttask` files

  Metadata from task and module files is merged into the project’s class-to-module map. Existing metadata is preserved when its providing module is unavailable, allowing projects and tasks containing dummy  objects to be saved without losing their module dependencies.

  Projects created before this change remain supported. Existing factory registration methods retain their original signatures and delegate to the new module-aware overloads.

Note: This PR is an initial work to support #1312 and #1305 

## How Has This Been Tested?

Focused unit tests were added for:

  - Class-to-module registration, lookup, removal, and duplicate handling
  - Backwards-compatible factory registration overloads
  - Project metadata loading and merging
  - Updating project mappings from live objects and existing metadata
  - Process-data loading
  - Task metadata serialization and preservation when modules are missing
  - Package metadata filtering
  - Externalized object metadata
  - XML reading, writing, filtering, legacy layout support, and malformed input

## Checklist:

- [x] A test for the new functionality was added (if applicable).
- [x] All tests run without failure.
- [ ] The changelog has been extended, if this MR contains important changes from the users's point of view.
- [x] The new code complies with the GTlab's style guide.
- [x] New interface methods / functions are exported via `EXPORT`. Non-interface functions are NOT exported.
- [x] The number of code quality warnings is not increasing.

## Summary by Sourcery

Introduce a class-to-module mapping manifest for projects and propagate these mappings through XML utilities, object factories, and project save/load, enabling retrieval of the providing module for each class used in a project.

New Features:
- Add XML support for reading and writing a CLASS-MODULES manifest and querying distinct object class names in project documents.
- Track and expose class-to-module associations in GtProject, including lookup of a class’s providing module.
- Extend object and task/calculator factories to register classes together with their providing module identifiers and to query a module by class name.

Enhancements:
- Ensure linked object files contain a filtered class-to-module manifest matching only the classes present in each linked file.
- Propagate class-to-module mappings when saving project XML and linked object files so module information is preserved across project I/O.

Tests:
- Add unit tests covering class-module manifest round-trip behavior, object class name discovery, project class-module mapping loading, linked file manifest filtering, and object factory module-id registration and lookup.

## Summary by Sourcery

Introduce project-wide class-to-module metadata and propagate it through XML utilities, object factories, and project/task save-load flows.

New Features:
- Add XML support for reading, writing, and filtering class-to-module manifests and querying distinct object class names in documents.
- Track class-to-module mappings in GtProject and expose lookup of the providing module for a given class.
- Extend object, calculator, and task factories to register classes alongside their providing module identifiers and query module IDs by class name.
- Persist class-provider metadata in task files and module package files, including preservation when referenced modules or classes are missing.

Enhancements:
- Populate and update project class-to-module mappings from loaded XML, task files, live objects, and module data.
- Ensure linked object and module files contain filtered class-to-module manifests limited to the classes they serialize.
- Include class-provider metadata when saving project XML, linked object files, and task group data while preserving unrelated metadata sections.

Tests:
- Add unit tests for XML class-module manifest round-trips, legacy layout support, and filtering by serialized objects.
- Add tests for factory class-module registration semantics, including duplicates, bulk registration, and module ID resolution through factory groups.
- Add project tests for loading, merging, and updating class-module mappings from project files, tasks, and modules, plus process-data loading behavior.
- Add task group tests for saving/loading class-provider metadata and preserving mappings for missing task and calculator classes.

### CODE REVIEW

File: src/core/process_management/gt_taskgroup.cpp [hunk 1]
```diff
@@ -28,6 +28,46 @@
 
 static const char* S_INDEX_FILE_NAME = "index.json";
 static const char* S_TASK_FILE_EXT = ".gttask";
+static const char* S_CLASS_PROVIDERS_PROPERTY = "_gt_classProviders";
+
+namespace
+{
+
+QVariantMap toVariantMap(const gt::xml::ClassModuleMap& mappings)
```
Comment (rainman110) [inline src/core/process_management/gt_taskgroup.cpp]:
```suggestion
QVariantMap
toVariantMap(const gt::xml::ClassModuleMap& mappings)
```

File: src/core/process_management/gt_taskgroup.cpp [hunk 2]
```diff
@@ -28,6 +28,46 @@
...
+{
+    QVariantMap result;
+    for (auto it = mappings.cbegin(); it != mappings.cend(); ++it)
+    {
+        result.insert(it.key(), it.value());
+    }
+    return result;
+}
+
+gt::xml::ClassModuleMap fromVariantMap(const QVariantMap& mappings)
```
Comment (rainman110) [inline src/core/process_management/gt_taskgroup.cpp]:
```suggestion
gt::xml::ClassModuleMap
fromVariantMap(const QVariantMap& mappings)
```

File: src/core/process_management/gt_taskgroup.cpp [hunk 3]
```diff
@@ -28,6 +28,46 @@
...
+    for (auto it = mappings.cbegin(); it != mappings.cend(); ++it)
+    {
+        result.insert(it.key(), it.value().toString());
+    }
+    return result;
+}
+
+void addClassProviderMetadata(
+    QDomDocument& doc, const gt::xml::ClassModuleMap& preserved = {})
+{
```
Comment (rainman110) [inline src/core/process_management/gt_taskgroup.cpp]:
```suggestion
void
addClassProviderMetadata(QDomDocument& doc, const gt::xml::ClassModuleMap& preserved = {})
{
```

File: src/core/gt_project.cpp [hunk 4]
```diff
@@ -45,6 +45,33 @@
 
 #include <cassert>
 
+namespace
+{
+
+gt::xml::ClassModuleMap taskClassModuleIds(const QString& projectPath)
```
Comment (rainman110) [inline src/core/gt_project.cpp]:
```suggestion
gt::xml::ClassModuleMap
taskClassModuleIds(const QString& projectPath)
```

File: src/dataprocessor/gt_abstractobjectfactory.h [hunk 5]
```diff
@@ -73,6 +73,14 @@ class GT_DATAMODEL_EXPORT GtAbstractObjectFactory
      */
     virtual bool registerClass(QMetaObject metaObj);
```
Comment (rainman110) [inline src/dataprocessor/gt_abstractobjectfactory.h:74]: we might need to deprecate this?

File: src/core/process_management/gt_taskgroup.cpp (hunk 1 above)
Comment (rainman110) [reply src/core/process_management/gt_taskgroup.cpp]: Applied in 9bdf13cf. Thanks.

File: src/core/process_management/gt_taskgroup.cpp (hunk 2 above)
Comment (rainman110) [reply src/core/process_management/gt_taskgroup.cpp]: Applied in 9bdf13cf. Thanks.

File: src/core/process_management/gt_taskgroup.cpp (hunk 3 above)
Comment (rainman110) [reply src/core/process_management/gt_taskgroup.cpp]: Applied in 9bdf13cf. Thanks.

File: src/core/gt_project.cpp (hunk 4 above)
Comment (rainman110) [reply src/core/gt_project.cpp]: Applied in 9bdf13cf. Thanks.

File: src/dataprocessor/gt_abstractobjectfactory.h (hunk 5 above)
Comment (rainman110) [reply src/dataprocessor/gt_abstractobjectfactory.h:74]: I’d prefer not to deprecate this overload. It is intentionally retained for source/API compatibility and remains valid for callers without module context; the framework module loaders now use the new module-aware overload. Deprecating it would create warnings for otherwise supported integrations.

File: src/dataprocessor/gt_xmlutilities.cpp [hunk 6]
```diff
@@ -79,6 +79,164 @@ sanitizeUuid(const QString& uuid)
...
+    QDomElement metadata =
+        root.firstChildElement(QLatin1String(S_METADATA_TAG));
+    QDomElement providers =
+        metadata.firstChildElement(QLatin1String(S_CLASS_PROVIDERS_TAG));
+    if (providers.isNull())
+    {
+        const QDomElement modules =
+            root.firstChildElement(QLatin1String(S_MODULES_TAG));
+        providers =
+            modules.firstChildElement(QLatin1String(S_CLASS_PROVIDERS_TAG));
```
[the remark this answers is not in the document]
Comment (rainman110) [reply src/dataprocessor/gt_xmlutilities.cpp:95]: Fixed in f03075af. `writeClassModuleMap()` now removes `MODULES/CLASS-PROVIDERS` before writing or clearing the modern metadata, so the document has one authoritative location. I also added regression coverage for rewriting a document that contains the legacy location.

File: src/core/process_management/gt_taskgroup.cpp [hunk 7]
```diff
@@ -28,6 +28,46 @@
...
+    QDomElement root = doc.documentElement();
+    gt::xml::ClassModuleMap mappings;
+    for (const QString& className : gt::xml::objectClassNames(root))
+    {
+        QString moduleId = gtProcessFactory->moduleId(className);
+        if (moduleId.isEmpty()) moduleId = preserved.value(className);
+        if (!moduleId.isEmpty()) mappings.insert(className, moduleId);
```
[the remark this answers is not in the document]
Comment (rainman110) [reply src/core/process_management/gt_taskgroup.cpp]: Fixed in f03075af. Task metadata lookup now checks `gtProcessFactory`, then `gtObjectFactory`, and finally preserved mappings. I omitted the suggested null guard because `gtObjectFactory` is the project singleton macro. A regression test covers a non-process object embedded in a task.

File: tests/unittests/core/test_gt_taskgroup.cpp [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    ASSERT_TRUE(taskFile.open(QIODevice::ReadOnly | QIODevice::Text));
+    QDomDocument doc;
+    ASSERT_TRUE(doc.setContent(&taskFile));
+    const auto mappings =
+        gt::xml::readClassModuleMap(doc.documentElement());
+    EXPECT_EQ(mappings.value(QStringLiteral("MetadataTestCalculator")),
+              QStringLiteral("CalculatorModule"));
+}
+
+TEST_F(GtTaskGroupMetadataTest, KeepsFileForMissingTaskClass)
```
[the remark this answers is not in the document]
Comment (rainman110) [reply tests/unittests/core/test_gt_taskgroup.cpp:172]: Added in f03075af. `KeepsFileForMissingTaskClass` now captures the task UUID, reloads and saves the task as a dummy, reads the generated `index.json`, and verifies that the UUID remains in the `active` array.

File: tests/unittests/datamodel/test_xmlutilities.cpp [hunk 9]
```diff
@@ -143,6 +143,211 @@ TEST_F(TestXmlUtilities, addObjectListElement)
...
+{
+    QDomDocument doc;
+    ASSERT_TRUE(doc.setContent(QStringLiteral(R"(
+        <GTLAB><METADATA><CLASS-PROVIDERS/></METADATA></GTLAB>)")));
+    QDomElement root = doc.documentElement();
+    gt::xml::takeClassModuleMap(root);
+    EXPECT_TRUE(root.firstChildElement(QStringLiteral("METADATA")).isNull());
+}
+
+TEST(GtXmlUtilities, TakesLegacyClassModuleMapFromModules)
```
[the remark this answers is not in the document]
Comment (rainman110) [reply tests/unittests/datamodel/test_xmlutilities.cpp:299]: Added mixed-layout coverage in f03075af. The test intentionally defines `METADATA` as authoritative rather than merging legacy entries, because the legacy block may be stale. It also verifies that both provider nodes are removed and unrelated metadata is preserved.

File: src/core/gt_project.h
Comment (mariusalexander) [inline src/core/gt_project.h:39]: Would it make sense to pimpl `GtProject` @rainman110? (not in this PR though)

File: src/dataprocessor/gt_xmlutilities.h [hunk 10]
```diff
@@ -31,6 +32,39 @@ namespace gt
 namespace xml
 {
 
+using ClassModuleMap = QMap<QString, QString>;
```
Comment (mariusalexander) [inline src/dataprocessor/gt_xmlutilities.h:36]: You could think about making this map more explicit (if I got this one right), alternatively maybe add a comment about what the key and what the value is representing?
```suggestion
using ClassName = QString;
using ModuleID = Qstring;
using ClassModuleMap = QMap<ClassName, ModuleID>;
```
OR
```suggestion
/// Maps class names to module ids
using ClassModuleMap = QMap<QString, QString>;
```

**Review by mariusalexander — CHANGES_REQUESTED** [review]

File: src/dataprocessor/gt_xmlutilities.h (hunk 10 above)
Comment (rainman110) [reply src/dataprocessor/gt_xmlutilities.h:36]: Addressed in cf83782f. I added a concise API comment documenting that the map keys are class names and the values are the IDs of their providing modules. Thanks.

File: src/core/gt_project.h
Comment (rainman110) [reply src/core/gt_project.h:39]: Yes, I think a pimpl would make sense for `GtProject`, especially to keep future private-state changes out of the public class layout. I agree that this should be handled in a separate PR.

**Comment by rainman110** [comment]
@mariusalexander happy to get your review, when you are back :)

File: src/core/gt_project.h
Comment (mariusalexander) [reply src/core/gt_project.h:39]: Created #1517

**Review by mariusalexander — APPROVED** [review]
